contracts/DataSourceDelegate.sol

```diff
@@ -1,86 +1,241 @@
+import '@jbx-protocol-v2/contracts/libraries/JBFundingCycleMetadataResolver.sol';
+import '@jbx-protocol-v2/contracts/libraries/JBTokens.sol';
+import '@jbx-protocol-v2/contracts/structs/JBFundingCycle.sol';
 
-  uint32 = 120;
+import '@paulrberg/contracts/math/PRBMath.sol';
+import '@uniswap/v3-periphery/contracts/libraries/OracleLibrary.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/interfaces/callback/IUniswapV3SwapCallback.sol';
+
+contract DataSourceDelegate is IJBFundingCycleDataSource, IJBPayDelegate, IUniswapV3SwapCallback {
+  using JBFundingCycleMetadataResolver for JBFundingCycle;
+  event Test(uint256);
+  error unAuth();
+  error Slippage();
+
+  IWETH9 private constant weth = IWETH9(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+  IJBToken private constant jbx = IJBToken(0x3abF2A4f8452cCC2CF7b4C1e4663147600646f66);
+
+  IUniswapV3Pool private constant pool = IUniswapV3Pool(0x48598Ff1Cee7b4d31f8f9050C2bbAE98e17E6b17);
+
+  IJBPayoutRedemptionPaymentTerminal public immutable jbxTerminal;
+  IJBSingleTokenPaymentTerminalStore public immutable terminalStore;
+  IJBDirectory public immutable directory;
+  IJBFundingCycleStore public immutable fundingCycleStore;
+
+  uint32 private constant twapPeriod = 120; // ADD SETTERS, shouldn't be constant + setter for max twap/spot deviation
+  uint256 public constant projectId = 1;
+  uint256 public immutable reservedRate;
+  uint256 private _swapTokenCount;
+  uint256 private _issueTokenCount;
+
+  constructor(IJBPayoutRedemptionPaymentTerminal _jbxTerminal, uint256 _reservedRate) {
+    terminalStore = _jbxTerminal.store();
+    directory = _jbxTerminal.directory();
+    fundingCycleStore = _jbxTerminal.store().fundingCycleStore();
+    jbxTerminal = _jbxTerminal;
+    reservedRate = _reservedRate;
+  }
 
   function payParams(JBPayParamsData calldata _data)
     external
-    view
     override
     returns (
       uint256 weight,
       string memory memo,
       IJBPayDelegate delegate
     )
   {
-    return (0, _data.memo, IJBPayDelegate(address(this)));
+    // The terminal must be the jbx terminal.
+    if (_data.terminal != jbxTerminal) revert unAuth();
+
+    // Get the current funding cycle.
+    JBFundingCycle memory _currentFundingCycle = fundingCycleStore.currentOf(_data.projectId);
+
+    // The funding cycle's reserved rate must be 100%, otherwise proceed with the normal behavior.
+    if (_currentFundingCycle.reservedRate() != JBConstants.MAX_RESERVED_RATE)
+      return (_data.weight, _data.memo, IJBPayDelegate(address(0)));
+
+    // Get the token count that would be distributed by minting a new supply.
+    uint256 _mintTokenCountQuote = PRBMath.mulDiv(
+      _data.amount.value,
+      _currentFundingCycle.weight,
+      10**18
+    );
+
+    // Get the token count that would be distributed by swapping.
+    uint256 _swapTokenCountQuote = OracleLibrary.getQuoteAtTick(
```

Comment:
how gas intensive is the quoting procedure? I recall it being nontrivial on uni v3

anyways would it be a good idea to guard this with an if statement that checks if _data.amount.value > 0?

Reply:
Check the oracle library, this is based the theoretical price based on the twap (so not really expensive) -> I still need to have a better 'twap/spot divergence' check in the callback (see discord)